### DISCUSSION TITLE
Readiness Discussion: SDK `1.0`

### DISCUSSION THREAD

**Discussion by aaronsteers** [meltano/sdk#962] [General]
### Context

The SDK is now [used in over 250 repos](https://github.com/meltano/sdk/network/dependents?package_id=UGFja2FnZS0xOTM3MjUxMDk3) and is [nearing 50 unique contributors](https://github.com/meltano/sdk/graphs/contributors).

<img width="289" alt="image" src="https://user-images.githubusercontent.com/18150651/190003234-c60367e1-cdbf-4021-88dc-6c957c57d9b8.png">

To the @meltano/meltano-team and to our contributors: 

1. Is there anything critical that we are missing feature-wise?
2. Any breaking changes or refactoring we'd want to do before dropping the `v1.0` release tag?

--------------

As originally discussed in [GitLab](https://gitlab.com/groups/meltano/-/epics/121).

[GitHub v1.0 Milestone](https://github.com/meltano/sdk/milestone/2)

**Comment by aaronsteers** [comment]
Pulling from [GitLab](https://gitlab.com/groups/meltano/-/epics/121) and more recent discussions:

- [ ] Make sure `context` arg is present in all applicable methods.
- [ ] Rename `rows` to `records`, esp. in function signatures: https://github.com/meltano/sdk/pull/966
- [ ] (Optional) Support `LOG_BASED` replication for DB-type sources: #304 
- [ ] (Optional) Support the ACTIVATE_MESSAGE unofficial singer extension: #18, #392

**Comment by edgarrmondragon** [comment]
Some things I would consider blockers for a _stable_ release:

- ~https://github.com/meltano/sdk/issues/208~
- https://github.com/meltano/sdk/issues/928

Other "nice to have", maybe breaking, changes:

- https://github.com/meltano/sdk/issues/933
- Not yet logged, but consider current `--about --format=markdown` vs simpler `--about=markdown`

Developer QOL:

- https://github.com/meltano/sdk/issues/877

**Comment by sbalnojan** [comment]
There might have already been a discussion about this, and I don't know the history, but how about **"the name"**?

What I mean:  Inside the docs there is currently only one place which tells me that it's the "SDK for taps and targets", so maybe if we start to use something like tt-sdk it becomes clearer what I can do with it?

Or better the dsdk (SDK for moving data) because the stream maps are also developed using the SDK as far as I can tell.

After all, Meltano lives from being customized & extended, and we need to get people pointed into the right direction right off the bat.

**Reply by edgarrmondragon** [reply]
@sbalnojan we probably want to still communicate that it's used to create Singer-compatible taps and targets, so maybe `meltano-singer-sdk`? Or, if we plan to go well beyond Singer maybe `meltano-flow-sdk`?

**Reply by sbalnojan** [reply]
@edgarrmondragon I get that point, true, we need to communicate that. I would just try to decouple it. I would prefer something like "meltano-flow-sdk" (I like that), and put a huge sign right next to it "built on the open standard Singer". 

Because this is the only thing that enables people to fully use Meltano. Every single "production ready" Meltano user will need to build 1-2 custom taps or targets, so there's no way around it. And everyone evaluating Meltano as a solution will ask "can I build custom taps/targets?", so I think clear communication is crucial here.

**Comment by kgpayne** [comment]
Things I would ❤️ to see before v1.0:

- Test framework & comprehensive standard tests ([in progress](https://github.com/meltano/sdk/pull/1171))
- Tap/Target versioning, at least for MeltanoLabs plugins ([in progress](https://github.com/MeltanoLabs/Meta/issues/22))
- SDK docs revamp, in line with @sbalnojans' excellent work on the Meltano docs
- [Refactor connectors](https://github.com/meltano/sdk/issues/317) out of Stream/Sink classes (resolves use cases where Taps need to authenticate to perform discovery and also mocking connections for Tap/Target testing)

**Comment by tayloramurphy** [comment]
@aaronsteers we should probably include this issue for v1

* https://github.com/meltano/sdk/issues/1409

**Comment by kgpayne** [comment]
@edgarrmondragon suggested in Office Hours:

- making SQLAlchemy optional (breaking change)

**Reply by edgarrmondragon** [reply]
Also `cryptography`, which we only use for the JWT authenticator.

**Comment by edgarrmondragon** [comment]
We should also make sure all breaking changes are documented in https://sdk.meltano.com/en/latest/deprecation.html

**Comment by kgpayne** [comment]
Another refactor that would be worth while would be to pull the `BATCH` file handling methods off of the Stream class onto a standalone `BatchIO` class (similar to the connections conversation). I ran into a [case](https://github.com/MeltanoLabs/target-snowflake/blob/main/target_snowflake/sinks.py#L355) with `target-snowflake` where I wanted to reuse the batch methods available to a `Stream` in my `Sink` class 🤔
